File: app/src/main/java/co/k2/newsbits/headlines/HeadlineViewModel.java
package co.k2.newsbits.headlines;

import android.content.Context;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.lifecycle.ViewModel;
import co.k2.newsbits.data.models.Article;
import co.k2.newsbits.data.source.NewsRepository;
import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

/**
 * Copyright (C) 2019 K2 CODEWORKS
 * All rights reserved
 *
 * @author Kislay
 * @since 29/03/19
 */

public class HeadlineViewModel extends ViewModel {

    private final NewsRepository newsRepository;
    private final Navigator navigator;

    HeadlineViewModel(NewsRepository repository, Navigator navigator) {
        this.newsRepository = repository;
        this.navigator = navigator;
    }

    Single<List<Article>> fetchNewArticles(String country) {
        return newsRepository.fetchNews(country)
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    Single<List<Article>> fetchCachedArticles() {
        return newsRepository.getCachedNews()
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread());
    }

    Disposable updateCache(List<Article> articles) {
        return newsRepository.clearCache()
                .andThen(newsRepository.saveToCache(articles))
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe();
    }

    void openNewsArticle(@NonNull Context context, @NonNull Article article) {
        navigator.navigateToWebPage(context, article.getUrl());
    }

}

File: app/src/main/java/co/k2/newsbits/headlines/HeadlineViewModel.kt
package co.k2.newsbits.headlines

import androidx.compose.runtime.mutableStateOf
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import co.k2.newsbits.common.isLessThanADayOld
import co.k2.newsbits.data.models.Article
import co.k2.newsbits.data.source.NewsRepository
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

/**
 * Copyright (C) 2019 K2 CODEWORKS
 * All rights reserved
 *
 * @author Kislay
 * @since 29/03/19
 */

class HeadlineViewModel constructor(
    private val newsRepository: NewsRepository
) : ViewModel() {

    val headlines = mutableStateOf<List<Article>?>(null)
    val error = mutableStateOf<Pair<Boolean, String?>>(Pair(false, null))

    fun fetchNewArticles(country: String) = viewModelScope.launch(Dispatchers.IO) {
        error.value = Pair(false, null)
        headlines.value = null
        val result = newsRepository.fetchNews(country)
        result.body?.articles?.let {
            headlines.value = it
            updateCache(it, country)
        } ?: run {
            fetchCachedArticles(country, result.error?.message)
        }
    }

    private fun fetchCachedArticles(
        country: String, errorMessage: String?
    ) = viewModelScope.launch(Dispatchers.IO) {
        val cacheInfo = newsRepository.getCacheMetadata()
        val lastRefresh = cacheInfo?.lastRefreshTime ?: 0L
        if (lastRefresh.isLessThanADayOld && country == cacheInfo?.country) {
            headlines.value = newsRepository.cachedNews()
            error.value = Pair(true, null)
        } else {
            error.value = Pair(true, errorMessage)
        }
    }

    private suspend fun updateCache(
        articles: List<Article>, country: String
    ) = withContext(Dispatchers.IO) {
        newsRepository.clearCache()
        newsRepository.saveToCache(articles, System.currentTimeMillis(), country)
    }

}
